examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py

```diff
@@ -0,0 +1,595 @@
+        # `equation = abc,cde->abde`, where `a`: batch size, `b`: sequence
+        # length, `c`: `hidden_dim`, `d`: `num_heads`,
+        # `e`: `hidden_dim//num_heads`. The only difference is that in layer `B`,
+        # `c` represents `rank`.
+        self.B = keras.layers.EinsumDense(
+            equation=original_layer_config["equation"],
+            output_shape=original_layer_config["output_shape"],
+            kernel_initializer="zeros",
+            trainable=trainable,
+            name=f"B_{layer_idx}",
+        )
+
+    def call(self, inputs):
+        original_output = self.original_layer(inputs)
+        if self.trainable:
+            # If we are fine-tuning the model, we will add LoRA layers' output
+            # to the original layer's output.
+            lora_output = self.B(self.A(inputs)) * self._scale
+            return original_output + lora_output
+
+        # If we are in inference mode, we "merge" the LoRA layers' weights into
+        # the original layer's weights - more on this in the text generation
+        # section!
+        return original_output
+
+
+"""
+### Inject LoRA layer into the model
+
+We will now hack the original GPT-2 model and inject LoRA layers into it. Let's
+do a couple of things before doing that:
+
+- Delete previous model;
+- Reset "peak" GPU memory usage using `tf.config.experimental.reset_memory_stats`;
+- Load a new GPT-2 model.
+"""
+
+del gpt2_lm
+del optimizer
+del loss
+
+# This resets "peak" memory usage to "current" memory usage.
+tf.config.experimental.reset_memory_stats("GPU:0")
+
+# Load the original model.
+preprocessor = keras_nlp.models.GPT2CausalLMPreprocessor.from_preset(
+    "gpt2_base_en",
+    sequence_length=128,
+)
+lora_model = keras_nlp.models.GPT2CausalLM.from_preset(
+    "gpt2_base_en",
+    preprocessor=preprocessor,
+)
+
+"""
+We will now override the original query/value projection matrices with our
+new LoRA layers.
+"""
+
+for layer_idx in range(lora_model.backbone.num_layers):
+    # Change query dense layer.
+    decoder_layer = lora_model.backbone.get_layer(f"transformer_layer_{layer_idx}")
+    self_attention_layer = decoder_layer._self_attention_layer
+
+    # Change query dense layer.
+    self_attention_layer._query_dense = LoraLayer(
+        self_attention_layer._query_dense,
+        rank=RANK,
+        alpha=ALPHA,
+        layer_idx=layer_idx,
+        trainable=True,
+    )
+
+    # Change value dense layer.
+    self_attention_layer._value_dense = LoraLayer(
+        self_attention_layer._value_dense,
+        rank=RANK,
+        alpha=ALPHA,
+        layer_idx=layer_idx,
+        trainable=True,
+    )
+
+"""
+Let's now do a forward pass to make sure we still have a valid chain of
+computation.
+"""
+
+lora_model(preprocessor(["they are going to ban LoRA in EU, lol"])[0])
+pass
+
+"""
+Freeze the entire LLM, only the LoRA layers should be trainable.
+"""
+
+for l in lora_model._flatten_layers():
```

Comment:
Prefer using fully spelled out names everywhere, e.g. `layer` instead of `l`, for readability